pyro/ops/ssm_gp.py

```diff
@@ -0,0 +1,139 @@
+import math
+
+import torch
+import torch.nn as nn
+
+
+root_three = math.sqrt(3.0)
+root_five = math.sqrt(5.0)
+five_thirds = 5.0 / 3.0
+
+
+class MaternKernel(nn.Module):
+    """
+    Provides the building blocks for representing univariate Gaussian Processes (GPs)
+    with Matern kernels as state space models.
+
+    :param float nu: The order of the Matern kernel (1.5 or 2.5)
+    :param int num_gps: the number of GPs
+    :param torch.tensor log_length_scale_init: optional `num_gps`-dimensional vector of initializers
+        for the length scale
+    :param torch.tensor log_kernel_scale_init: optional `num_gps`-dimensional vector of initializers
+        for the kernel scale
+
+    **References**
+
+    [1] `Kalman Filtering and Smoothing Solutions to Temporal Gaussian Process Regression Models`,
+        Jouni Hartikainen and Simo Sarkka.
+    [2] `Stochastic Differential Equation Methods for Spatio-Temporal Gaussian Process Regression`,
+        Arno Solin.
+    """
+    def __init__(self, nu=1.5, num_gps=1, log_length_scale_init=None, log_kernel_scale_init=None):
+        assert nu in [1.5, 2.5], "The only supported values of nu are 1.5 and 2.5"
+        self.nu = nu
+        self.state_dim = {1.5: 2, 2.5: 3}[nu]
+        self.num_gps = num_gps
+
+        if log_length_scale_init is not None:
```

Comment:
nit: in general it is cleaner to locate assertions after control flow so it is obvious that both code paths follow the same invariant. While you can indeed figure that out by reading the code, `sed` isn't as smart and often makes mistakes when editing code.
```py
if log_length_scale is None:
    log_length_scale_init = torch.zeros(num_gps)
assert log_length_scale_init.shape == (num_gps,)

if log_kernel_scale_init is None:
    log_kernel_scale_init = torch.zeros(num_gps)
assert log_kernel_scale_init.shape == (num_gps,)
```

Reply:
done

Reply:
done